apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs

```diff
@@ -46,6 +108,196 @@ private SpannerConversionOptions(bool useDBNull)
+        }
+
+        internal SpannerConversionOptions WithClrToSpannerMappings(string clrToSpannerMappings)
+        {
+            var clone = Clone();
+            // We already have a new dictionary being returned, so no need to use Dictionary constructor.
+            clone.ClrToSpannerTypeMappings = ValidateAndParseClrToSpannerTypeMappings(clrToSpannerMappings);
+            return clone;
+        }
+
+        internal SpannerConversionOptions WithSpannerToClrMappings(string spannerToClrMappings)
+        {
+            var clone = Clone();
+            // We already have a new dictionary being returned, so no need to use Dictionary constructor.
+            clone.SpannerToClrTypeMappings = ValidateAndParseSpannerToClrTypeMappings(spannerToClrMappings);
+            return clone;
+        }
+
+        private ConcurrentDictionary<SpannerDbType, System.Type> ValidateAndParseSpannerToClrTypeMappings(string input)
+        {
+            // Empty string or null is valid to unset the values.
+            if (string.IsNullOrEmpty(input))
+            {
+                // return defaults.
+                return DefaultSpannerToClrTypeMappings();
+            }
+
+            // Get all type mappings.
+            var mappings = input.Split(new char[] { ',', ';' }, StringSplitOptions.RemoveEmptyEntries);
+
+            // Validate that mappings are valid.
+            if (mappings == null || !mappings.Any())
+            {
+                throw new ArgumentException($"'{input}' is not a valid value for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // Validate that mapping matches one of the pre-defined values.
+            var invalidMappings = mappings.Except(AllowedSpannerToClrTypeMappings, StringComparer.OrdinalIgnoreCase);
+            if (invalidMappings.Any())
+            {
+                throw new ArgumentException($"The following mappings are invalid:'{string.Join(",", invalidMappings)}' for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // Check multiples for each type.
+            // Currently, we have Float64 and Date to check.
+            if (mappings.Count(j => j.StartsWith("Float64", StringComparison.OrdinalIgnoreCase)) > 1)
+            {
+                throw new ArgumentException($"'{input}' is not a valid value as multiple mappings from Float64 to CLR type are provided for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            if (mappings.Count(j => j.StartsWith("Date", StringComparison.OrdinalIgnoreCase)) > 1)
+            {
+                throw new ArgumentException($"'{input}' is not a valid value as multiple mappings from Date to CLR type are provided for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // We don't want to modify the current instance, so create a new dictionary.
```

Comment:
I suspect we can do this without a dictionary to be honest... but now the logic is in the "right" place, I'm happier to refactor later.